### ISSUE TITLE
salt-minion hangs when receiving SIGTERM (solaris, possible everywhere)

### ISSUE THREAD

**Issue by sjorge** [saltstack/salt#32146]
### Description of Issue/Question
sending a SIGTERM to a salt-minion to have it shutdown results in stuck processes:
```
root     69056  4.4  0.1 71668 8372 ?        W 16:42:35  0:09 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     69060  4.4  0.1 71668 7632 ?        O 16:42:35  0:09 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     69059  4.3  0.1 71668 7656 ?        W 16:42:35  0:09 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     69062  4.3  0.1 71668 7652 ?        R 16:42:35  0:09 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     69055  4.3  0.1 71668 8340 ?        R 16:42:35  0:09 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     69058  4.3  0.1 71668 7664 ?        O 16:42:35  0:09 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     69061  4.2  0.1 71668 7588 ?        R 16:42:35  0:09 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     69092  2.5  0.1 71668 7280 ?        W 16:43:05  0:02 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     69091  2.5  0.1 71668 8140 ?        R 16:43:05  0:03 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     69089  2.4  0.1 71668 7292 ?        R 16:43:05  0:02 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     69057  2.3  0.1 71668 8356 ?        S 16:42:35  0:07 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     69063  2.3  0.1 71668 7612 ?        S 16:42:35  0:07 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     69080  1.9  0.1 71668 8144 ?        O 16:43:05  0:02 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     69031  1.1  0.1 71668 7832 ?        S 16:42:20  0:11 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     69029  1.1  0.1 71668 8564 ?        S 16:42:20  0:11 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     69030  1.1  0.1 71668 7840 ?        S 16:42:20  0:11 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     69028  0.4  0.2 71412 11404 ?        S 16:42:13  0:06 python2.7 /opt/salt/bin/salt-minion -c /salt/config
```

This is what an idle minion looks like:
```
root     68041  5.1  1.0 99684 78884 ?        S 16:37:03  0:06 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     67958  1.6  0.5 63816 37984 ?        S 16:36:53  0:02 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     67943  0.3  0.4 40352 31244 ?        S 16:36:52  0:00 python2.7 /opt/salt/bin/salt-minion -c /salt/config
root     67985  0.0  0.1 46244 3616 ?        S 16:36:56  0:00 python2.7 /opt/salt/bin/salt-minion -c /salt/config
```

Setting 'multiprocessing' to false fixes this behavior. I'm not sure how much running with threads is tested vs processes. Master seems to lack the support to run threaded so I am assuming it not as well tested.

### Setup
n/a just running salt-minion and sending a SIGTERM

### Steps to Reproduce Issue
I can't reproduce this when running a salt-minion and sending it a SIGTERM, it only happens when it has an active connection open. e.g. when the service is restarted using 'salt x service.restart salt-minion'

### Versions Report
```yaml
Salt Version:
           Salt: 2016.3.0rc1-112-g1af2e41

Dependency Versions:
         Jinja2: 2.8
       M2Crypto: 0.22
           Mako: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.4.1
         Python: 2.7.11 (default, Mar 18 2016, 13:38:08)
           RAET: 0.6.5
        Tornado: 4.3
            ZMQ: 4.1.3
           cffi: Not Installed
       cherrypy: 3.8.0
       dateutil: 2.4.0
          gitdb: 0.6.4
      gitpython: 1.0.2
          ioflo: 1.5.1
        libgit2: Not Installed
        libnacl: 1.4.4
   msgpack-pure: Not Installed
 msgpack-python: 0.4.7
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: 2.0.2
          smmap: 0.9.0
        timelib: 0.2.4

System Versions:
           dist:
        machine: i86pc
        release: 5.11
         system: SunOS
        version: Not Installed
```

**Comment by sjorge** [comment]
From the minion log when it gets stuck:
```
The salt minion is shutdown. Minion received a SIGTERM. Exited.Exception SystemExit: 0 in <bound method AsyncZeroMQPubChannel.__del__ of <salt.transport.zeromq.AsyncZeroMQPubChannel object at 0xfffffd7ffdaeb790>> ignored
```

**Comment by sjorge** [comment]
Switched the solaris smf service to ':kill -KILL' for now, it... well rather harsh on the minion but atleast it gets restarted now.

**Comment by anlutro** [comment]
Do you have an event returner configured?

**Comment by sjorge** [comment]
Not exactly sure what that is. Is one configured by default?

**Comment by anlutro** [comment]
Nah. Was just wondering if it was related to https://github.com/saltstack/salt/issues/31575 but I guess not. I also somehow missed that you were talking about a minion, not a master, so ignore me!

**Comment by sjorge** [comment]
Yeah just the minion, master stops fine with just a TERM

**Comment by jfindlay** [comment]
@sjorge, thanks for reporting.

**Comment by sjorge** [comment]
```
[WARNING ] Minion received a SIGTERM. Exiting.
[INFO    ] The salt minion is shutting down..
[INFO    ] The Salt Minion is shut down
The salt minion is shutdown. Minion received a SIGTERM. Exited.#
```

All fixed in 2016.3.3
